bitcoin/script/miniscript.h

```diff
@@ -348,6 +349,106 @@ struct Node {
+         * State variables are omitted for simplicity.
+         *
+         * First: stack=[(A,false)] results=[].
+         *        stack=[(A,true),(F,false),(C,false),(B,false)] results=[]
+         *        stack=[(A,true),(F,false),(C,false)] results=[B]
+         *        stack=[(A,true),(F,false),(C,true),(E,false),(D,false)] results=[B]
+         *        stack=[(A,true),(F,false),(C,true),(E,false)] results=[B,D]
+         *        stack=[(A,true),(F,false),(C,true)] results=[B,D,E]
+         *        stack=[(A,true),(F,false)] results=[B,C]
+         *        stack=[(A,true)] results=[B,C,F]
+         * Final: stack=[] results=[A]
+         */
+        while (stack.size()) {
+            const Node& node = stack.back().node;
+            if (!stack.back().expanded && !node.subs.empty()) {
+                /* We encounter an unexpanded tree node with children. Mark it as
+                 * expanded, and add its children in reverse order. By the time we
+                 * process this node again, the results of the children will be
+                 * at the end of results in normal order. */
+                stack.back().expanded = true;
+                stack.reserve(stack.size() + node.subs.size());
+                const auto& parent_state = stack.back().state;
+                for (size_t i = 0; i < node.subs.size(); ++i) {
+                    size_t index = node.subs.size() - i - 1;
+                    stack.emplace_back(**(node.subs.rbegin() + i), false,
+                        downfn(parent_state, node, index));
+                }
+                continue;
+            }
+            // Invoke upfn with the last node.subs.size() elements of results as input.
+            assert(results.size() >= node.subs.size());
+            std::optional<Result> result{upfn(std::move(stack.back().state), node,
+                Span{results}.last(node.subs.size()))};
+            // If evaluation returns std::nullopt, abort immediately.
+            if (!result) return {};
+            // Replace the last node.subs.size() elements of results with the new result.
+            results.resize(results.size() - node.subs.size() + 1);
+            results.back() = std::move(*result);
+            stack.pop_back();
+        }
+        // The final remaining results element is the root result, return it.
+        assert(results.size() == 1);
+        return std::move(results[0]);
+    }
+
+    /** Like TreeEvalMaybe, but always produces a result. upfn must return Result. */
+    template<typename Result, typename State, typename DownFn, typename UpFn>
+    Result TreeEval(State root_state, DownFn&& downfn, UpFn upfn) const
+    {
+        // Invoke TreeEvalMaybe with upfn wrapped to return std::optional<Result>, and then
+        // unconditionally dereference the result (it cannot be std::nullopt).
+        return std::move(*TreeEvalMaybe(std::move(root_state), std::forward<downfn>(downfn),
```

Comment:
Even after applying the following fix this doesn't compile:
```suggestion
        return std::move(*TreeEvalMaybe(std::move(root_state), std::forward<DownFn>(downfn),
```
```c++
In file included from script/descriptor.cpp:9:
./script/miniscript.h: In instantiation of ‘Result miniscript::Node<Key>::TreeEval(State, DownFn&&, UpFn) const [with Result = CScript; State = bool; DownFn = miniscript::Node<Key>::ToScript<{anonymous}::ScriptMaker>::<lambda(bool, const miniscript::Node<unsigned int>&, size_t)>&; UpFn = miniscript::Node<Key>::ToScript<{anonymous}::ScriptMaker>::<lambda(bool, const miniscript::Node<unsigned int>&, Span<CScript>)>; Key = unsigned int]’:
./script/miniscript.h:531:33:   required from ‘CScript miniscript::Node<Key>::ToScript(const Ctx&) const [with Ctx = {anonymous}::ScriptMaker; Key = unsigned int]’
script/descriptor.cpp:929:57:   required from here
./script/miniscript.h:444:40: error: no matching function for call to ‘miniscript::Node<unsigned int>::TreeEvalMaybe(std::remove_reference<bool&>::type, miniscript::Node<Key>::ToScript<{anonymous}::ScriptMaker>::<lambda(bool, const miniscript::Node<unsigned int>&, size_t)>&, miniscript::Node<Key>::TreeEval<CScript, bool, miniscript::Node<Key>::ToScript<{anonymous}::ScriptMaker>::<lambda(bool, const miniscript::Node<unsigned int>&, size_t)>&, miniscript::Node<Key>::ToScript<{anonymous}::ScriptMaker>::<lambda(bool, const miniscript::Node<unsigned int>&, Span<CScript>)> >::<lambda(bool&&, const miniscript::Node<unsigned int>&, Span<CScript>)>) const’
  444 |         return std::move(*TreeEvalMaybe(std::move(root_state), std::forward<DownFn>(downfn),
      |                           ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  445 |             [&upfn](State&& state, const Node& node, Span<Result> subs) {
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  446 |                 Result res{upfn(std::move(state), node, subs)};
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  447 |                 return std::optional<Result>(std::move(res));
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  448 |             }
      |             ~                           
  449 |         ));
      |         ~                               
./script/miniscript.h:373:27: note: candidate: ‘template<class Result, class State, class DownFn, class UpFn> std::optional<_Up> miniscript::Node<Key>::TreeEvalMaybe(State, DownFn, UpFn) const [with Result = Result; State = State; DownFn = DownFn; UpFn = UpFn; Key = unsigned int]’
  373 |     std::optional<Result> TreeEvalMaybe(State root_state, DownFn downfn, UpFn upfn) const
      |                           ^~~~~~~~~~~~~
./script/miniscript.h:373:27: note:   template argument deduction/substitution failed:
./script/miniscript.h:444:40: note:   couldn’t deduce template parameter ‘Result’
  444 |         return std::move(*TreeEvalMaybe(std::move(root_state), std::forward<DownFn>(downfn),
      |                           ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  445 |             [&upfn](State&& state, const Node& node, Span<Result> subs) {
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  446 |                 Result res{upfn(std::move(state), node, subs)};
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  447 |                 return std::optional<Result>(std::move(res));
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  448 |             }
      |             ~                           
  449 |         ));
      |         ~                               
make[2]: *** [Makefile:8966: script/libbitcoin_common_a-descriptor.o] Error 1
make[1]: *** [Makefile:17338: all-recursive] Error 1
make: *** [Makefile:822: all-recursive] Error 1
```
And i'm admittedly a bit out of my depth on the template deduction here.

Reply:
Fixed.